packages/mds-policy-service/tests/validators.spec.ts

```diff
@@ -436,4 +437,145 @@ describe('Tests Policy Intents', () => {
       expect(validatePermittedParkingPolicyIntent(validIntent)).not.toBeFalsy()
     })
   })
+
+  describe('Test ProviderRebalancingZone intent', () => {
+    it('passes a valid intent', () => {
+      const validIntent = {
+        intent_type: 'provider_rebalancing_zone',
+        policy_fields: DEFAULT_POLICY_USER_FIELDS,
+        rule_fields: [
+          { ...DEFAULT_RULE_FIELDS, maximum: 100, states: {} },
+          { ...DEFAULT_RULE_FIELDS, maximum: 50, states: {} },
+          {
+            ...DEFAULT_RULE_FIELDS,
+            geographies: [uuid(), uuid()],
+            maximum: 0,
+            states: {
+              available: ['provider_drop_off'],
+              elsewhere: ['provider_drop_off'],
+              non_operational: ['provider_drop_off'],
+              on_trip: ['provider_drop_off'],
+              removed: ['provider_drop_off'],
+              reserved: ['provider_drop_off'],
+              unknown: ['provider_drop_off']
+            }
+          }
+        ]
+      }
+
+      expect(validateProviderRebalancingZonePolicyIntent(validIntent)).not.toBeFalsy()
+    })
+
+    it('throws if the states for the last rule are wrong', () => {
+      const invalidIntent = {
+        intent_type: 'provider_rebalancing_zone',
+        policy_fields: DEFAULT_POLICY_USER_FIELDS,
+        rule_fields: [
+          { ...DEFAULT_RULE_FIELDS, maximum: 100, states: {} },
+          { ...DEFAULT_RULE_FIELDS, maximum: 50, states: {} },
+          {
+            ...DEFAULT_RULE_FIELDS,
+            maximum: 100,
+            states: {
+              available: ['provider_drop_off'],
+              elsewhere: ['provider_drop_off'],
+              non_operational: ['provider_drop_off'],
+              on_trip: ['provider_drop_off'],
+              removed: ['provider_drop_off'],
+              reserved: ['provider_drop_off'],
+              unknown: ['provider_drop_off']
+            }
+          }
+        ]
+      }
+
+      expect(() => validateProviderRebalancingZonePolicyIntent(invalidIntent)).toThrow(ValidationError)
```

Comment:
I'd match a regex against the validation error message to make sure that the expected validation failure is being caught. In this instance the test states that the states are wrong and we're expecting the validator to throw an error on states, but actually it's the maximum field that's wrong here and results in the validator throwing. Explicitly matching against the error message would help catch these types of issues where multiple different kinds of errors might be thrown.

Reply:
Sadly I couldn't figure how to get AJV to throw a nicer error than "validation for custom keyword failed" (not the exact wording of the actual AJV error, but that's about what I got). It'll require further investigating into the AJV api to figure how to get detailed error messages like "maximum < 0, this is incorrect."